link/src/main/java/com/stripe/android/link/ui/verification/VerificationViewModel.kt

```diff
@@ -0,0 +1,220 @@
+import com.stripe.android.core.strings.resolvableString
+import com.stripe.android.link.LinkActivityResult
+import com.stripe.android.link.LinkScreen
+import com.stripe.android.link.account.LinkAccountManager
+import com.stripe.android.link.analytics.LinkEventsReporter
+import com.stripe.android.link.injection.NativeLinkComponent
+import com.stripe.android.link.model.AccountStatus
+import com.stripe.android.link.ui.ErrorMessage
+import com.stripe.android.link.ui.getErrorMessage
+import com.stripe.android.ui.core.elements.OTPSpec
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+/**
+ * ViewModel that handles user verification confirmation logic.
+ */
+internal class VerificationViewModel @Inject constructor(
+    private val linkAccountManager: LinkAccountManager,
+    private val linkEventsReporter: LinkEventsReporter,
+    private val logger: Logger,
+    private val goBack: () -> Unit,
+    private val navigateAndClearStack: (route: LinkScreen) -> Unit,
+    private val dismissWithResult: (LinkActivityResult) -> Unit
+) : ViewModel() {
+
+    private val _viewState = MutableStateFlow(
+        value = run {
+            val linkAccount = linkAccountManager.linkAccount.value
+            if (linkAccount == null) {
+                dismissWithResult(LinkActivityResult.Failed(NoLinkAccountFoundForVerification()))
+            }
+            VerificationViewState(
+                redactedPhoneNumber = linkAccount?.redactedPhoneNumber ?: "",
+                email = linkAccount?.email ?: ""
+            )
+        }
+    )
+    val viewState: StateFlow<VerificationViewState> = _viewState
+
+    val otpElement = OTPSpec.transform()
+
+    private val otpCode: StateFlow<String?> =
+        otpElement.otpCompleteFlow.stateIn(viewModelScope, SharingStarted.Lazily, null)
+
+    init {
+        setUp()
+    }
+
+    private fun setUp() {
+        val linkAccount = linkAccountManager.linkAccount.value
+
+        if (linkAccount == null) {
+            dismissWithResult(LinkActivityResult.Failed(NoLinkAccountFoundForVerification()))
+            return
+        }
+        if (linkAccount.accountStatus != AccountStatus.VerificationStarted) {
+            startVerification()
+        }
+
+        linkEventsReporter.on2FAStart()
+
+        viewModelScope.launch {
+            otpCode.collect { code ->
+                code?.let { onVerificationCodeEntered(code) }
+            }
+        }
+    }
+
+    fun onVerificationCodeEntered(code: String) {
+        updateViewState {
+            it.copy(
+                isProcessing = true,
+                errorMessage = null,
+            )
+        }
+
+        viewModelScope.launch {
+            linkAccountManager.confirmVerification(code).fold(
```

Comment:
I think having a launch here means that we effectively have some code like:
```
viewModelScope.launch {
    otpCode.collect {
         viewModelScope.launch { ...
```
Should we instead just have the second launch also be on the top-level? That seems better than having them nested (from a clarity perspective and also efficiency)